### PR TITLE
Implement a macOS listener

### PR DESCRIPTION
**Pull request by zwimer**
Implements: https://github.com/albertosottile/darkdetect/issues/25

UPDATED: This PR:
1. Adds a macOS listener; this listener requires `pip install darkdetect[mac_listener]`; if the extra `mac_listener` is not installed, the listener raises `NotImpelementedError`
2. Bumps minor version to `0.8.0` as a feature has been added.
3. Drops python2 support, as per: https://github.com/albertosottile/darkdetect/pull/28#issuecomment-1339961348

### CODE REVIEW

**Comment by albertosottile** [comment]
Upon further investigation, it appears that also `PyObjCTools.AppHelper.runConsoleEventLoop` returns immediately when calling from a secondary thread. I guess we are then back to the problems reported in https://github.com/albertosottile/darkdetect/issues/25#issuecomment-1188297721.

**Comment by zwimer** [comment]
@albertosottile Concerning extras, I could, but I will point out that the way I added it to `requires` is such that it only gets downloaded on `macOS` and only for `python3.0+`. With those edits in place, I feel this should stay as required rather than an extra, as this is part of the core functionality advertised on the README. If users really want to ignore dependencies via pip they could always do `pip install --no-deps darkdetect`. If you disagree with this though, let me know.

As for the `README.md` update: It works in the main thread, I'm having trouble getting it to work in a worker thread- I'll look into this.

**Comment by zwimer** [comment]
This PR now requires python3, which I've taken is ok given: https://github.com/albertosottile/darkdetect/pull/28#issuecomment-1339961348

Right now it uses `multiprocessing`, instead of `subprocess` which may have side effects if a user forgets `if __name__ == '__main__':` and their code is run in `spawn` instead of `fork` mode (as in, the child process may re-import things / the main file's code). It could be edited to use `subprocess`, but that would require mimicking an IPC queue; I think this is preferring multiprocessing is beneficial.

**Comment by zwimer** [comment]
I'll update the PR description with all the changes.

**Comment by albertosottile** [comment]
I am sorry for all the effort you are investing in this, perhaps I should have been clearer with the requirements. The goal of `listener` is to provide a function that continuously monitors the GUI theme and calls a callback function whenever there are changes. It is then left to the user how to run this, whether in the main thread, or in a separate thread, or in a separate process or subprocess. This was the original idea behind the `listener` API, and is also part of the "contract" reported in the README. To be honest, I would prefer to have no macOS listener, than a listener that does not allow to fulfill this contract.

In addition to this genera principle, I would especially avoid using `multiprocessing` in `darkdetect`, as it could have complex and unpredictable effects on the users' code. You mention forgetting `if __name__ == '__main__'`, but this is just the tip of the iceberg. The truth is that `multiprocessing` completely reinitializes all the import chain, and this can cause issues with a lot of GUI libraries (e.g. `PySide2` and `wx`) that the user can simply not control, and `darkdetect` is actually meant to be used in conjunction with such libraries.

In summary, I would rather not merge the PR as it is. I will adapt the text in #25 to clarify the requirements for this function.

**Comment by albertosottile** [comment]
> To be honest, I would prefer to have no macOS listener, than a listener that does not allow to fulfill this contract.

As further clarification, I would like to emphasize that most GUI bindings actually provide native methods to continuously detect theme changes, see e.g. an implementation for Qt here: https://github.com/albertosottile/darkdetect/issues/14#issuecomment-945090847 

(the same is not true for determining dark vs light, hence the need for `darkdetect`).

**Comment by zwimer** [comment]
I can adapt this to use subprocess then that waits on and reads stdout; just like how the linux listener works. That should satisfy every requirement listed:
```python
def listener(callback):
    with subprocess.Popen(
        ('gsettings', 'monitor', 'org.gnome.desktop.interface', 'gtk-theme'),
        stdout=subprocess.PIPE,
        universal_newlines=True,
    ) as p:
        for line in p.stdout:
            callback('Dark' if '-dark' in line.strip().removeprefix("gtk-theme: '").removesuffix("'").lower() else 'Light')

```
Except, instead of `('gsettings', 'monitor', 'org.gnome.desktop.interface', 'gtk-theme')`, it'll be `('sys.executable', '-c', 'import darkdetect; darkdetect._macos_listener._listen_child()')` or something similar.

As for QT, that is actually the reason I've found this problem, since that signal was insufficient for determining if the theme changed when used with custom palletes. Unfortunately we've found no native way to truly detect light vs dark mode in QT.

**Comment by zwimer** [comment]
The latest commit updates the listener to be basically as the above, just like the linux listener.

**Comment by zwimer** [comment]
Rebased onto master because of the `pyproject.toml` update.

**Comment by albertosottile** [comment]
> I can adapt this to use subprocess then that waits on and reads stdout; just like how the linux listener works. That should satisfy every requirement listed:

I am fine with this approach, provided that it fulfills all the requirements listed in #25.

> As for QT, that is actually the reason I've found this problem, since that signal was insufficient for determining if the theme changed when used with custom palletes. Unfortunately we've found no native way to truly detect light vs dark mode in QT.

True, but Qt can notify you when there is a change, then you can call `darkdetect.theme()` to determine the new appearance.

> The latest commit removes multiprocessing and updates the listener to be basically as the above, just like the linux listener.

I tried your implementation on my MacBook Air, but it does not work, not even in the main thread: nothing is printed when I switch from Dark to Light and vice versa. Did you test this code?

File: darkdetect/_mac_detect.py [hunk 1]
```diff
@@ -69,6 +80,38 @@ def isDark():
...
+    if not _can_listen:
+        raise NotImplementedError()
+    pth = f"import sys; sys.path.insert(0, r'''{Path(__file__).parents[1]}''')"
+    sig = "import signal as s; s.signal(s.SIGINT, s.SIG_IGN)"
+    listen = "import darkdetect as dd; dd._mac_detect._listen_child()"
+    with subprocess.Popen(
+        (sys.executable, "-c", f"{pth}; {sig}; {listen}"),
+        stdout=subprocess.PIPE,
+        # stderr=subprocess.DEVNULL,
+        universal_newlines=True,
```
Comment (albertosottile) [inline darkdetect/_mac_detect.py:114]: I would avoid a `sys.path` manipulation and instead use the `cwd` argument in `subprocess.Popen`. Suggested change:
```python
    cwd = Path(__file__).parent
    listen = "import _mac_detect as md; md._listen_child()"
    with subprocess.Popen(
        (sys.executable, "-c", f"{listen}"),
        stdout=subprocess.PIPE,
        universal_newlines=True,
        cwd=cwd,
```

File: pyproject.toml [hunk 2]
```diff
@@ -35,6 +34,9 @@ email = "asottile@gmail.com"
 homepage = "http://github.com/albertosottile/darkdetect"
 download = "http://github.com/albertosottile/darkdetect/releases"
 
+[project.optional-dependencies]
+mac_listener = [ "pyobjc-framework-Cocoa; platform_system == 'Darwin'" ]
```
Comment (albertosottile) [inline pyproject.toml:38]: I would rename the extra as `macos-listener`.

File: darkdetect/_mac_detect.py [hunk 3]
```diff
@@ -69,6 +80,38 @@ def isDark():
...
+            except IOError:
+                os._exit(1)
+    observer = Observer.new()  # Keep a reference of the observer to keep it alive
+    NSDistributedNotificationCenter.defaultCenter().addObserver_selector_name_object_(
+        observer,  # Observer must be kept alive by a different reference
+        "callback:",
+        "AppleInterfaceThemeChangedNotification",
+        None
+    )
+    AppHelper.runConsoleEventLoop()
```
Comment (albertosottile) [inline darkdetect/_mac_detect.py:101]: Currently, this code does not work for me. If you want, I can provide another implementation based on `AppleInterfaceStyle` that works with the rest of the PR.

**Comment by zwimer** [comment]
> Did you test this

I did. What code did you use to test this? I can run it myself

```python
(test3) zwimer@Lotus ~/D/W/darkdetect> python
Python 3.10.8 (main, Oct 13 2022, 09:48:40) [Clang 14.0.0 (clang-1400.0.29.102)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import threading
>>> import darkdetect
>>>
>>> # def listener(callback: typing.Callable[[str], None]) -> None: ...
>>>
>>> t = threading.Thread(target=darkdetect.listener, args=(print,))
>>> t.daemon = True
>>> t.start()
>>> Dark
Light
Dark
Light
[1]    21169 quit       python3
```

**Comment by albertosottile** [comment]
Same code as you, but also `darkdetect.listener(print)` from the main thread: the callback is never called when I change theme. I am on macOS 12.6.1

File: darkdetect/_mac_detect.py (hunk 3 above)
Comment (zwimer) [reply darkdetect/_mac_detect.py:101]: If it doesn't work for you, then I'll try your alternative observer.

**Comment by zwimer** [comment]
Hmm, if you send over your alternative I shall try integrating it.

File: darkdetect/_mac_detect.py (hunk 3 above)
Comment (albertosottile) [reply darkdetect/_mac_detect.py:101]: Import: `from Foundation import NSObject, NSKeyValueObservingOptionNew, NSKeyValueChangeNewKey, NSUserDefaults`

```python
def _listen_child():
    OBSERVED_KEY = "AppleInterfaceStyle"

    class Observer(NSObject):
        def observeValueForKeyPath_ofObject_change_context_(
            self, path, object, changeDescription, context
        ):
            result = changeDescription[NSKeyValueChangeNewKey]
            if result is None:
                result = "Light"
            print(f'{result}', flush=True)

    observer = Observer.new()
    defaults = NSUserDefaults.standardUserDefaults()

    defaults.addObserver_forKeyPath_options_context_(
        observer, OBSERVED_KEY, NSKeyValueObservingOptionNew, 0
    )

    AppHelper.runConsoleEventLoop(installInterrupt=True) 
```

**Comment by zwimer** [comment]
The last commit should address the requested changes.

File: README.md [hunk 4]
```diff
@@ -49,6 +49,13 @@ pip install darkdetect
 
 Alternatively, you are free to vendor directly a copy of Darkdetect in your app. Further information on vendoring can be found [here](https://medium.com/underdog-io-engineering/vendoring-python-dependencies-with-pip-b9eb6078b9c0).
 
+## Optional Installs
+
+To enable the macOS listener, additional components are required, these can be installed via:
+```bash
+pip install -e darkdetect[macos-listener]
```
Comment (albertosottile) [inline README.md:56]: `-e` is not necessary here.

File: darkdetect/_mac_detect.py [hunk 5]
```diff
@@ -69,6 +82,37 @@ def isDark():
...
+
+
+def listener(callback: Callable[[str], None]) -> None:
+    if not _can_listen:
+        raise NotImplementedError()
+    sig = "import signal as s; s.signal(s.SIGINT, s.SIG_IGN)"
```
Comment (albertosottile) [inline darkdetect/_mac_detect.py:109]: Is this truly necessary? With my implementation, I was able to interrupt the subprocess even without it:
```python
^CTraceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/alberto/darkdetect_zwimer/darkdetect/_mac_detect.py", line 132, in listener
    for line in p.stdout:
KeyboardInterrupt
```

Perhaps the key is to use `AppHelper.runConsoleEventLoop(installInterrupt=True)`.

File: darkdetect/_mac_detect.py [hunk 6]
```diff
@@ -69,6 +82,37 @@ def isDark():
...
+    listen = "import darkdetect as dd; dd._mac_detect._listen_child()"
```
Comment (albertosottile) [inline darkdetect/_mac_detect.py:110]: In this case, I would avoid importing `darkdetect` as this could use a system-installed copy. As suggested before, I would import `_mac_detect` directly and set `cwd=Path(__file__).parent` to avoid this potential ambiguity.

**Review by albertosottile — CHANGES_REQUESTED** [review]

File: darkdetect/_mac_detect.py (hunk 5 above)
Comment (zwimer) [reply darkdetect/_mac_detect.py:109]: The reason I did this is to fix the problem where: the main program catches `KeyboardInterrupt` and the listener shouldn't die as a result. Depending if you are in a terminal or not, this signal may end up being sent to the listener process, which we probably do not want.

It also has a bonus of ensuring no log spam when the listener dies for a KeyboardInterrupt, but that is a side effect, not the reason.

File: darkdetect/_mac_detect.py (hunk 6 above)
Comment (zwimer) [reply darkdetect/_mac_detect.py:110]: I see, I misunderstood you earlier. I'll try that!

**Comment by zwimer** [comment]
> True, but Qt can notify you when there is a change, then you can call darkdetect.theme() to determine the new appearance.

Unfortunately for my application, since the pallete is customizeable it can be manually set so this wouldn't work either. This is a failing on QT's part, other frameworks do implement theme detection notifications. That's why I intend to fall back to darkdetect.

File: darkdetect/_mac_detect.py (hunk 3 above)
Comment (zwimer) [reply darkdetect/_mac_detect.py:103]: I'll give it a shot.

File: darkdetect/_mac_detect.py (hunk 5 above)
Comment (albertosottile) [reply darkdetect/_mac_detect.py:109]: I honestly do not see any difference with my implementation, but this should not be harmful. Another possibility is to add the signal ignore code in the beginning of `_listen_child`.

**Comment by zwimer** [comment]
@albertosottile That should use the listener you provided. I've tested it on my end; let me know if it works for you.

**Comment by albertosottile** [comment]
> That should use the listener you provided. I've tested it on my end; let me know if it works for you.

Works perfectly both from the main and from a secondary thread. Thanks!

File: darkdetect/_mac_detect.py (hunk 5 above)
Comment (albertosottile) [reply darkdetect/_mac_detect.py:111]: For me the PR would be complete as this is not actually a requirement. However, because the signal addition does not seem to do anything, I would remove it from the code.

File: darkdetect/_mac_detect.py [hunk 7]
```diff
@@ -6,6 +6,19 @@
 
 import ctypes
 import ctypes.util
+import subprocess
+import sys
+import os
```
Comment (albertosottile) [inline darkdetect/_mac_detect.py:11]: This import is no longer needed.

Comment (zwimer) [reply darkdetect/_mac_detect.py:11]: I'll remove the `os` one; the other two are for the main listener function.

File: darkdetect/_mac_detect.py (hunk 5 above)
Comment (zwimer) [reply darkdetect/_mac_detect.py]: This was caused by a bug; I've fixed it.

**Comment by zwimer** [comment]
The last commit:
1. Properly ignores Ctrl-C
2. On death, if the parent doesn't clean up the child, the child will be silent when it later dies.

**Review by albertosottile — APPROVED** [review]
LGTM, I tested it with my `KeyboardInterrupt` script and it worked perfectly.

**Comment by albertosottile** [comment]
Thanks for having worked on this. I will make a release with this feature when I have some spare time.
